### ISSUE TITLE
Adding label support to the new compiler and runtime

### ISSUE THREAD

**Issue by wodeni** [penrose/penrose#111]
The recent change in #99 did not include support for labels as of 180717. We should add the functionalities from #94 into the new system. The proposed steps are the following:

- `collectLabels :: [String] -> [SubStmt] -> LabelMap` is the top-level label resolution function that expects (1) a list of all substance identifiers and (2) a list of substance statements
- After parsing Substance, a `type LabelMap = M.Map String (Maybe String)` is created separately.
- `translateStyProg` will be modified to take in a `LabelMap` in addition to the old parameters. 
- After all the translation is done, call a new function `insertLabels :: labelMap -> Translation -> Translation` to insert the labels. Specifically:
  - __Question: should the label of a GPI be rendered if there is no `Text { }` defined explicitly?__
  - Assuming the question is __yes__, then we loop through all keys in `trMap` and look for corresponding label, and then create a `Text` GPI as a field (named `text`). 
  - All the rest of the pipeline will just run as normal

**Comment by hypotext** [comment]
Example file we discussed:

```
Set A {
    -- equivalent to 
    -- A.label = "\int_1^100..."

    -- NoLabel A
    -- A.label = ""

    A.shape = Circle {
    }   

    A.shape = Bracket {
    }   
}

Set A; Set B
where Labeled A; NoLabel B {
      -- A.label

      override A.shape.x = computeSomething(A.text)

    A.text = Text { 
    	   string = check(A.label) ++ "hi"
    }
    -- A.text.string, A.text

    A.left_label = Text {
    	   string = check(A.label) ++ " left"
    }

    A.labelFn = encourage near(A.left_label, A.zzz) 
    -- what does A.shape.text refer to? the GPI or the string?
}

-- where NoLabel A -- Label A x

-- Alternative to selector matching on label/nolabel:
-- have the Style compiler invisibly decide not to create/use the text GPI AND anything that uses the text GPI

-- Another approach: NoLabel A -> A.label = "" -> "check" function makes sure that all GPIs setting the string fo a text using a label get the ENTIRE empty string
```

**Comment by wodeni** [comment]
Current solution
* Each Substance entry (NOT Style entry like `Color`) in translation map will have a field called `label`, which stores the __lebel text__ (without `$$`).
* If there is no label specified for a Substance object because of (1) `NoLabel` (2) the lack of `Label` and/or `AutoLabel` statements, `label` field will be an __empty string__
* Since we have no conditionals on whether an object has a label, we don't have a very principle solution to deal with the absence of labels in the Style program. 
  * __Temporary__:  In the front-end, we explicitly ignore `Text` GPIs with an empty `text` property during rendering

Tested on the following input by commenting one label line at a time
```
Scalar a
Scalar c
Scalar m

AutoLabel a, c
Label a $\Delta$
Label c $\beta$
NoLabel a
NoLabel c
```

**Comment by hypotext** [comment]
Decisions:

- Labels are built into the DSLL's prelude
- We want to be able to match on labels or no labels
- What is `Autolabel`?

**Comment by wodeni** [comment]
Solved in #94. Closing the issue now.
